ultravox/tools/ds_tool/ds_tool.py

```diff
@@ -223,6 +228,124 @@ def __post_init__(self):
             self.upload_split = self.dataset_split
 
 
+class DatasetChunkProcessor:
+    args: DatasetToolArgs
+    cache_dir: str = ".cache/ds_tool/processed_datasets"
+    chunks_not_uploaded: List[Tuple[int, int]] = []
+    total_samples_processed: Dict[str, int] = {}
+
+    def __init__(self, args: DatasetToolArgs):
+        self.args = args
+
+    def process_and_upload_split(self, split_name: str, ds_split: datasets.Dataset):
+        failed_chunk_ranges = self._dynamic_chunk(
+            split_name, ds_split, 0, len(ds_split)
+        )
+
+        while len(failed_chunk_ranges) > 0:
+            new_failed_ranges = []
+            for start, end in failed_chunk_ranges:
+                print(f"Retrying failed chunk range [{start}, {end})")
+                new_failed_ranges.extend(
+                    self._dynamic_chunk(split_name, ds_split, start, end)
+                )
```

Comment:
Seems like you can have a dataset where only a portion of the samples are uploaded, is that true?

I need to read more of the code, but that seems wrong.
1. Does the order of samples change?
3. Is having a half-finished dataset okay? What was the trade-off, e.g. why waiting to process then upload (with retry) doesn't work?

Reply:
1. The order of samples will change if there's a failure. But it doesn't seem important to me to keep the order?
2. It won't be half-finished, will most likely be 99% finished, with the 1% unfinished being the problematic part of the dataset (ie, ill formatted data, random Null values that the filter didn't catch).